demos/induction_heads_multigpu.py

```diff
@@ -0,0 +1,285 @@
+    MixedPrecision,
+    ShardingStrategy,
+)
+from torch.distributed.fsdp.wrap import (
+    transformer_auto_wrap_policy,
+)
+from transformers import (
+    LlamaConfig,
+    LlamaForCausalLM,
+)
+from transformers.models.llama.modeling_llama import LlamaDecoderLayer
+
+from flex_model.core import FlexModel, HookFunction
+
+
+def setup() -> None:
+    """Instantiate process group."""
+    dist.init_process_group("nccl")
+
+
+def cleanup() -> None:
+    """Destroy process group."""
+    dist.destroy_process_group()
+
+
+def args() -> Namespace:
+    """Parse command-line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_path", type=str, required=True)
+    return parser.parse_args()
+
+
+def setup_model(model_path: str, rank: int) -> tuple[nn.Module, LlamaConfig]:
+    """Instantiate model, tokenizer, and config.
+
+    Args:
+    ----
+        model_path: A path to the model being instantiated
+        rank: The worker rank
+
+    Returns:
+    -------
+        A tuple of length two containing the model and the config.
+    """
+    config = LlamaConfig.from_pretrained(model_path)
+    if rank == 0:
+        model = LlamaForCausalLM.from_pretrained(
+            model_path, torch_dtype=torch.bfloat16,
+        )
+    else:
+        with torch.device("meta"):
+            model = LlamaForCausalLM.from_pretrained(
+                model_path, torch_dtype=torch.bfloat16,
+            )
+    return model, config
+
+
+def fsdp_config(rank: int) -> dict[str: Any]:
+    """Return the config to be used by FSDP.
+
+    Args:
+    ----
+        rank: The worker rank
+
+    Returns:
+    -------
+        A dictionary containing keyword -> respective configuration.
+    """
+    def _module_init_fn(module: nn.Module) -> Callable:
+        """Return the function used for initializing modules on FSDP workers."""
+        return module.to_empty(
+            device=torch.cuda.current_device(),
+            recurse=False,
+        )
+
+    auto_wrap_policy = partial(
+        transformer_auto_wrap_policy,
+        transformer_layer_cls={
+            LlamaDecoderLayer,
+        },
+    )
+    sharding_strategy = ShardingStrategy.FULL_SHARD
+    device_id = torch.cuda.current_device()
+    sync_module_states = True
+    param_init_fn = _module_init_fn if rank != 0 else None
+    mp_policy = MixedPrecision(
+        param_dtype=torch.bfloat16,
+        buffer_dtype=torch.bfloat16,
+        reduce_dtype=torch.bfloat16,
+    )
+
+    config = {
+        "auto_wrap_policy": auto_wrap_policy,
+        "sharding_strategy": sharding_strategy,
+        "device_id": device_id,
+        "sync_module_states": sync_module_states,
+        "param_init_fn": param_init_fn,
+        "mixed_precision": mp_policy,
+    }
+    return config
+
+
+def calculate_induction_score(
+        config: LlamaConfig,
+        activation_dict: dict[str, torch.Tensor],
+        module_names: list[str],
+        sequence_length: int,
```

Comment:
Would probably be better to just pass the `num_hidden_layers` and `num_attention_heads` rather than the entire `LlamaConfig`, since you're already passing the `sequence_length`